plugins/in_splunk/splunk_prot.c

```diff
@@ -0,0 +1,822 @@
+    if (header->key.data == NULL) {
+        send_response(conn, 400, "error: header 'Content-Type' is not set\n");
+        return -1;
+    }
+
+    if (header->val.len == 10 &&
+        strncasecmp(header->val.data, "text/plain", 10) == 0) {
+        type = HTTP_CONTENT_TEXT;
+    }
+
+    if (type == -1) {
+        /* Not neccesary to specify content-type for Splunk HEC. */
+        flb_plg_debug(ctx->ins, "Mark as unknown type for ingested payloads");
+        type = HTTP_CONTENT_UNKNOWN;
+    }
+
+    if (request->data.len <= 0) {
+        send_response(conn, 400, "error: no payload found\n");
+        return -1;
+    }
+
+    /* Always handle as raw type of payloads here */
+    if (type == HTTP_CONTENT_TEXT || type == HTTP_CONTENT_UNKNOWN) {
+        process_raw_payload_pack(ctx, tag, request->data.data, request->data.len);
+    }
+
+    return 0;
+}
+
+static inline int mk_http_point_header(mk_ptr_t *h,
+                                       struct mk_http_parser *parser, int key)
+{
+    struct mk_http_header *header;
+
+    header = &parser->headers[key];
+    if (header->type == key) {
+        h->data = header->val.data;
+        h->len  = header->val.len;
+        return 0;
+    }
+    else {
+        h->data = NULL;
+        h->len  = -1;
+    }
+
+    return -1;
+}
+
+/*
+ * Handle an incoming request. It perform extra checks over the request, if
+ * everything is OK, it enqueue the incoming payload.
+ */
+int splunk_prot_handle(struct flb_splunk *ctx, struct splunk_conn *conn,
+                       struct mk_http_session *session,
+                       struct mk_http_request *request)
+{
+    int i;
+    int ret;
+    int len;
+    char *uri;
+    char *qs;
+    off_t diff;
+    flb_sds_t tag;
+    struct mk_http_header *header;
+
+    if (request->uri.data[0] != '/') {
+        send_response(conn, 400, "error: invalid request\n");
+        return -1;
+    }
+
+    /* Decode URI */
+    uri = mk_utils_url_decode(request->uri);
+    if (!uri) {
+        uri = mk_mem_alloc_z(request->uri.len + 1);
+        if (!uri) {
+            return -1;
+        }
+        memcpy(uri, request->uri.data, request->uri.len);
+        uri[request->uri.len] = '\0';
+    }
+
+    /* Try to match a query string so we can remove it */
+    qs = strchr(uri, '?');
+    if (qs) {
+        /* remove the query string part */
+        diff = qs - uri;
+        uri[diff] = '\0';
+    }
+
+    /* Refer the tag at first*/
+    if (ctx->ins->tag && !ctx->ins->tag_default) {
+        tag = flb_sds_create_size(1024);
+        if (!tag) {
+            return -1;
+        }
+        /* clear tag before using */
+        flb_sds_len_set(tag, 0);
+        flb_sds_cat(tag, ctx->ins->tag, flb_sds_len(ctx->ins->tag));
+    }
+    else {
+        /* Compose the query string using the URI */
+        len = strlen(uri);
+
+        if (len == 1) {
+            tag = NULL; /* use default tag */
+        }
+        else {
+            tag = flb_sds_create_size(len);
```

Comment:
Can't we just call `flb_sds_create_len` with `&url[1]` and `len - 1` as arguments here and eliminate the need for line 704?

Reply:
Off course. We should use the simpler operation. I tested your suggestion and succeeded to handle converting from uri to tag operation.
Fixed in https://github.com/fluent/fluent-bit/pull/7349/commits/f7c73e04ca7eb2d86c47478bc519bae15cc8aa46.